File: GO/99.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

var first *TreeNode
var last *TreeNode
var pre *TreeNode
 
func helper(root *TreeNode) {
	if root.Left!=nil {
		helper(root.Left)
	}
	if pre!=nil {
		if root.Val < pre.Val {
			if first==nil {
				first=pre
			}
			last=root
		}
	}
	pre=root
	if root.Right!=nil {
		helper(root.Right)
	}
}
func recoverTree(root *TreeNode)  {
	first=nil
	last=nil
	pre=nil
	if root!=nil {
		helper(root)
		temp:=first.Val
		first.Val=last.Val
		last.Val=temp
	}
}

File: JAVA/99.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private TreeNode first,last,pre;
    private void helper(TreeNode root){
        if(root.left!=null){
            helper(root.left);
        }
        if(pre!=null){
            if(root.val<pre.val){
                if(first==null) first=pre;
                last=root;
            }
        }
        pre=root;
        if(root.right!=null){
            helper(root.right);
        }
    }
    public void recoverTree(TreeNode root) {
        first=null;
        last=null;
        pre=null;
        if(root!=null){
            helper(root);
            int temp=first.val;
            first.val=last.val;
            last.val=temp;
        }
    }
}